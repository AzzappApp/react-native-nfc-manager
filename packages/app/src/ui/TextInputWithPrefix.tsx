import { forwardRef, useState } from 'react';
import { TextInput as NativeTextInput, View, StyleSheet } from 'react-native';
import { colors, textStyles } from '#theme';
import { createStyleSheet, useStyleSheet } from '#helpers/createStyles';
import type { DefinedColorSchemeName } from '#helpers/createStyles';
import type { ForwardedRef } from 'react';
import type {
  TextInputProps as NativeTextInputProps,
  NativeSyntheticEvent,
  TextInputFocusEventData,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';

export type TextInputProps = Omit<NativeTextInputProps, 'style'> & {
  /**
   * Whether the input is in error state
   */
  isErrored?: boolean;
  /**
   * The main container style
   */
  style?: StyleProp<ViewStyle>;
  /**
   * The main container style
   */
  inputStyle?: StyleProp<TextStyle> | undefined;
  /**
   * input prefix
   */
  prefix?: string;
};

/**
 * A wrapper around TextInput that adds Azzapp's default styling.
 *
 */
const TextInputWithPrefix = (
  {
    isErrored,
    onFocus,
    onBlur,
    style,
    inputStyle,
    prefix,
    ...props
  }: TextInputProps,
  ref: ForwardedRef<NativeTextInput>,
) => {
  //#region hooks
  const styles = useStyleSheet(styleSheet);
  const [isFocused, setIsFocused] = useState(false);
  //#endregion

  const onFocusInner = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setIsFocused(true);
    onFocus?.(e);
  };

  const onBlurInner = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    setIsFocused(false);
    onBlur?.(e);
  };

  return (
    <View
      style={[
        styles.input,
        isFocused && styles.focused,
        isErrored && styles.errored,
        mainStyles.container,
        style,
      ]}
    >
      {prefix && <NativeTextInput value={prefix} style={styles.prefix} />}
      <NativeTextInput
        ref={ref}
        testID="nativeInputText"
        selectionColor={colors.primary400}
        style={[styles.text, inputStyle]}
        {...props}
        placeholderTextColor={colors.grey400}
        onFocus={onFocusInner}
        onBlur={onBlurInner}
        allowFontScaling={false}
      />
    </View>
  );
};

export const styleSheetData = (appearance: DefinedColorSchemeName) => ({
  text: {
    flex: 1,
    color: appearance === 'light' ? colors.black : colors.white,
  },
  input: {
    ...textStyles.textField,
    paddingHorizontal: 15,
    paddingVertical: 3,
    height: 47,
    backgroundColor: appearance === 'light' ? colors.grey50 : colors.grey1000,
    borderWidth: 1,
    borderColor: appearance === 'light' ? colors.grey50 : colors.grey1000,
    borderRadius: 12,
    color: appearance === 'light' ? colors.black : colors.white,
  },
  prefix: {
    minWidth: 30,
    color: colors.grey400,
    paddingRight: 5,
  },
  focused: {
    borderColor: appearance === 'light' ? colors.grey900 : colors.grey400,
  },
  errored: {
    borderColor: colors.red400,
  },
});
const styleSheet = createStyleSheet(styleSheetData);

const mainStyles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

export default forwardRef(TextInputWithPrefix);
