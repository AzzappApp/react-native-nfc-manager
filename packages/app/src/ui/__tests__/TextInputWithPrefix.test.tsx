import { fireEvent, render, screen } from '@testing-library/react-native';
import TextInputWithPrefix from '../TextInputWithPrefix';

const mockedUseColorScheme = jest.fn();

jest.mock('react-native/Libraries/Utilities/useColorScheme', () => {
  return {
    default: mockedUseColorScheme,
  };
});

describe('TextInputWithPrefix', () => {
  test('empty component', () => {
    render(<TextInputWithPrefix />);
    expect(screen).toMatchSnapshot();
  });
  test('component with value', () => {
    render(<TextInputWithPrefix value="test" />);
    expect(screen).toMatchSnapshot();
  });
  test('component with value and prefix', () => {
    render(<TextInputWithPrefix value="test" prefix="test" />);
    expect(screen).toMatchSnapshot();
  });
  test('component change text value', () => {
    const component = render(<TextInputWithPrefix value="test" />);
    const inputText = component.getByTestId('nativeInputText');
    fireEvent.changeText(inputText, 'text');
    expect(screen).toMatchSnapshot();
  });
});
