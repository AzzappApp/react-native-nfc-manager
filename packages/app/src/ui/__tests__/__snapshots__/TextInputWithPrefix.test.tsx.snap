// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputWithPrefix component change text value 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F5F5F6",
        "borderColor": "#F5F5F6",
        "borderRadius": 12,
        "borderWidth": 1,
        "color": "#0E1216",
        "fontFamily": "PlusJakartaSans-Regular",
        "fontSize": 16,
        "height": 47,
        "lineHeight": 20,
        "paddingHorizontal": 15,
        "paddingVertical": 3,
      },
      false,
      undefined,
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <TextInput
    allowFontScaling={false}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholderTextColor="#A1A1A5"
    selectionColor="#FF502E"
    style={
      [
        {
          "color": "#0E1216",
          "flex": 1,
        },
        undefined,
      ]
    }
    testID="nativeInputText"
    value="test"
  />
</View>
`;

exports[`TextInputWithPrefix component with value 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F5F5F6",
        "borderColor": "#F5F5F6",
        "borderRadius": 12,
        "borderWidth": 1,
        "color": "#0E1216",
        "fontFamily": "PlusJakartaSans-Regular",
        "fontSize": 16,
        "height": 47,
        "lineHeight": 20,
        "paddingHorizontal": 15,
        "paddingVertical": 3,
      },
      false,
      undefined,
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <TextInput
    allowFontScaling={false}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholderTextColor="#A1A1A5"
    selectionColor="#FF502E"
    style={
      [
        {
          "color": "#0E1216",
          "flex": 1,
        },
        undefined,
      ]
    }
    testID="nativeInputText"
    value="test"
  />
</View>
`;

exports[`TextInputWithPrefix component with value and prefix 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F5F5F6",
        "borderColor": "#F5F5F6",
        "borderRadius": 12,
        "borderWidth": 1,
        "color": "#0E1216",
        "fontFamily": "PlusJakartaSans-Regular",
        "fontSize": 16,
        "height": 47,
        "lineHeight": 20,
        "paddingHorizontal": 15,
        "paddingVertical": 3,
      },
      false,
      undefined,
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <TextInput
    style={
      {
        "color": "#A1A1A5",
        "minWidth": 30,
        "paddingRight": 5,
      }
    }
    value="test"
  />
  <TextInput
    allowFontScaling={false}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholderTextColor="#A1A1A5"
    selectionColor="#FF502E"
    style={
      [
        {
          "color": "#0E1216",
          "flex": 1,
        },
        undefined,
      ]
    }
    testID="nativeInputText"
    value="test"
  />
</View>
`;

exports[`TextInputWithPrefix empty component 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#F5F5F6",
        "borderColor": "#F5F5F6",
        "borderRadius": 12,
        "borderWidth": 1,
        "color": "#0E1216",
        "fontFamily": "PlusJakartaSans-Regular",
        "fontSize": 16,
        "height": 47,
        "lineHeight": 20,
        "paddingHorizontal": 15,
        "paddingVertical": 3,
      },
      false,
      undefined,
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <TextInput
    allowFontScaling={false}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholderTextColor="#A1A1A5"
    selectionColor="#FF502E"
    style={
      [
        {
          "color": "#0E1216",
          "flex": 1,
        },
        undefined,
      ]
    }
    testID="nativeInputText"
  />
</View>
`;
